agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+      zend_string_addref(key->filename);
+    }
+  }
+}
+
+void nr_php_wraprec_hashmap_key_release(nr_php_wraprec_hashmap_key_t* key) {
+  if (NULL != key->scope_name) {
+    zend_string_release(key->scope_name);
+  }
+  if (NULL != key->function_name) {
+    zend_string_release(key->function_name);
+  }
+  if (NULL != key->filename) {
+    zend_string_release(key->filename);
+  }
+  key->lineno = 0;
+}
+
+static inline size_t nr_zendfunc2bucketidx(size_t log2_num_buckets,
+                                           zend_function* zf) {
+  /* default to lineno */
+  uint32_t hash = nr_php_zend_function_lineno(zf);
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    /* but use hash of function name when possible */
+    hash = ZSTR_HASH(zf->op_array.function_name);
+  } else if (NULL != zf->op_array.filename) {
+    /* and as last resort try hash of filename if available */
+    hash = ZSTR_HASH(zf->op_array.filename);
+  }
+
+  /* normalize to stay in-bound */
+  return (hash & ((1 << log2_num_buckets) - 1));
+}
+
+static inline bool zstr_equal(zend_string* zs1, zend_string* zs2) {
+  if (NULL == zs1 || NULL == zs2) {
+    return false;
+  }
+
+  if (ZSTR_LEN(zs1) != ZSTR_LEN(zs2)) {
+    return false;
+  }
+
+  if (ZSTR_HASH(zs1) != ZSTR_HASH(zs2)) {
+    return false;
+  }
+
+  if (0 != memcmp(ZSTR_VAL(zs1), ZSTR_VAL(zs2), ZSTR_LEN(zs1))) {
+    return false;
+  }
+
+  return true;
+}
+
+static bool nr_is_wraprec_for_zend_func(nr_php_wraprec_hashmap_key_t* key,
+                                        zend_function* zf) {
+  /* Start with comparing line number: */
+  if (nr_php_zend_function_lineno(zf) != key->lineno) {
+    /* no match: line number is different - no need to check anything else */
+    return false;
+  }
+
+  /* Next compare function name unless it is unnamed closure. Zend engine sets
+   * function name to '{closure}' for all unnamed closures so function name
+   * cannot be used for them. A fallback method to compare filename is used
+   * for unnamed closures. */
+  if (zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    if (!zstr_equal(key->function_name, zf->op_array.function_name)) {
+      /* no match: function name is different - no need to check anything else
+       */
+      return false;
+    }
+    /* If function is scoped, compare the scope: */
+    if (zf->op_array.scope
+        && !zstr_equal(key->scope_name, zf->op_array.scope->name)) {
+      /* no match: scope is different */
+      return false;
+    }
+    /* match: line number, function name and scope (if function is scoped) are
+     * the same */
+    return true;
+  }
+
+  /* deal with unnamed colsure: fallback to comparing filename */
+  if (!zstr_equal(key->filename, zf->op_array.filename)) {
+    /* no match: filename is different */
+    return false;
+  }
+
+  /* match: line number and filename are the same */
+  return true;
+}
+
+static int nr_php_wraprec_hashmap_fetch_internal(
```

Comment:
this looks like a truthy function, is there a reason we don't want the return type to be `bool` rather than int?

Reply:
API copy/pasted from [axiom's hashmap implementation](https://github.com/misiekn/newrelic-php-agent/blob/refactor/wraprec-lookup/axiom/util_hashmap.c#L286-L305).